### ISSUE TITLE
Pool incompatibility

### ISSUE THREAD

**Issue by 2naive** [sidorares/node-mysql2#1078]
Hi there,

It was very helpful, that here:
https://github.com/mysqljs/mysql/blob/master/lib/Pool.js#L15
-- we have properties, that could be used as given:
`pool.on('connection', function (connection) {
    var p_all = pool._allConnections.length;
    var p_fre = pool._freeConnections.length;
    var p_ack = pool._acquiringConnections.length;
    var p_que = pool._connectionQueue.length;

    console.log('[MYSQL] Connection established\tConnections: %s\tACK: %s\tQUE: %s\tFRE:',
        p_all,
        p_ack,
        p_que,
        p_fre
    );
});`

When moved to mysql2 -- it breaks. Could it be restored?

Thanks!

**Comment by joeyhub** [comment]
Usually by convention anything with _ is meant to be private so you're not really meant to use it.

**Comment by sidorares** [comment]
@2naive are you talking about `connection` event or "private" pool methods? If latter, we'll need to agree with mysqljs/mysql on api and expose some public methods for this. As @joeyhub says, methods starting with `_` are private by convention

**Comment by 2naive** [comment]
@sidorares @joeyhub hi, I get your point in saying that "_params" meant to be private, but:
1. they are not, and are widely used, e.g. first found:
https://github.com/danmactough/node-mysql-wrapped/blob/38e49c34222beced3c70ac8a97e9643ef4a63f14/test/index.js#L141
https://github.com/ghostxbh/node_graphql/blob/1800bd1172b7d02dab631060b609a62e53f2beb5/utils/mysql-util.js#L28
https://github.com/EricHug/EBlog_end/blob/eaa53801ff7e9e59d3378f2e20ad5c0cdac49f96/database/db.js#L24
2. I have no clue, why they should be private at all
3. mysqljs/mysql dont provide us any getters for them
4. if we are talking about 100% compatibility, so now it's not

Thanks!

**Comment by sidorares** [comment]
1 - https://www.hyrumslaw.com/

**Comment by joeyhub** [comment]
@2naive it is unfortunately standard that all systems have internal and external interfaces with it being standard practice to mark internal with a _ prefix in systems that have limitations around making things private or encapsulated. Some people also allow use of internals by users "at their own risk".

You're either meant to only use the external ones or generally not be able to complain if when using internal interfaces you change version and it breaks something. Exceptions are rare and tend to be unintended.

It can happen that important features were only made for internal use with some significant use case for external use due to oversight, etc.

External interfaces tend to be more thought out carefully and usually frozen or change less, when they do change versioning and other mechanisms usually reflect this and help users to change what they have to. These interfaces are meant to be bridges between A and B so that A can change their internals and only having to worry about breaking compatibility if they change the external interface.

If you make it so you can't change the internal interface everything can get frozen, you can't change anything, can't fix bugs, improve how things work, etc.

If people are using internals though, which aren't meant to be used normally, then it suggests some kind of feature request to make an official external API for those things. Even though external interfaces can still be subject to change that change can then be catered to.

**Comment by 2naive** [comment]
Okay, guys, aside from holywar: how may I track connections: total, queue, freed?
@sidorares you proposed to make some public getter?

**Comment by 2naive** [comment]
For now I took just:
`pool._allConnections.toArray().length`
but what was the reason of removing _acquiringConnections?

And what's "the right way" of migratin with this:
```
    connection._socket.setKeepAlive(true, tcp_keepalive_timeout);

    connection._socket.setTimeout(tcp_idle_timeout);
    connection._socket.on('timeout', function() {
        // do stuff
        connection._socket.destroy();
        connection.destroy();
    });
```
?

Thanks!

**Comment by 2naive** [comment]
Looks like due to: https://github.com/sidorares/node-mysql2/issues/1098
connection.stream works well.

**Comment by sidorares** [comment]
> but what was the reason of removing _acquiringConnections?

I don't think it was removed, probably just some recent addition that was not added to mysql2. Pool implementation is shared between 2 libraries almost as is with very little modifications. If you can find a time to port latest pool to mysql2 @2naive that would be really appreciated

re public interface to monitor inner state of the pool - yes, we can add that. Would be good to list first possible use cases how they would be used so we can plan for possible edge cases and agree on good naming

Can you list your example - why do you need to know _allConnections / _acquiringConnections / _acquiringConnections / _connectionQueue ? Is it just for debugging purposes? Maybe adding some tracing using https://www.npmjs.com/package/debug would be enough?

**Comment by 2naive** [comment]
@sidorares 
> I don't think it was removed, probably just some recent addition that was not added to mysql2.
It was there at least from 2014: https://github.com/mysqljs/mysql/commit/76de66e2c36a6da71dcd010b3afe0029e5eb68bb#diff-1b589a05fcf1509343500faf9cd70bda

> Can you list your example 
I had some cases in production:
a) table lock
b) db's host machine stale
c) connection problems
d) zombie sockets
- this could lead to situation, e.g. when all active sockets are actually dead and ACK/QUE is rising.
In such cases I have to inspect and handle connections manually.
That's why I trace the state of pool connections and have some conditions to take place.

Indeed it could be used for debug purpose, but it's better to have in logs clean numbers with stats of connections for investigation of issue if needed, not whole details of what's was happening with sockets underline

Thanks
